### PR TITLE
Arbitrary date filter

### PR DESCRIPTION
**Pull request by calellowitz**
This creates a filter for mobile UCR that allows you to specify an arbitrary date range
spec here:
https://docs.google.com/document/d/1G6N0xaQAy7cgHyuGnX1DiL9E-uRkW208gyP2wftuhgI/edit#
screenshots:
datefilter1
datefilter2
@nickpell @orangejenny cc: @snopoke @biyeun

### CODE REVIEW

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -3340,6 +3341,53 @@ def get_filter_value(self, user):
...
+        elif self.operator == '=':
+            start_date = end_date = today - datetime.timedelta(days=days)
+        elif self.operator == '>=':
+            start_date = None
+            end_date = today - datetime.timedelta(days=days)
+        elif self.operator == '<=':
+            start_date = today - datetime.timedelta(days=days)
+            end_date = None
+        elif self.operator == '<':
+            start_date = today - datetime.timedelta(days=days-1)
```
Comment (dimagimon) [inline corehq/apps/app_manager/models.py]: E226 missing whitespace around arithmetic operator

File: corehq/apps/app_manager/models.py [hunk 2]
```diff
@@ -3340,6 +3341,53 @@ def get_filter_value(self, user):
...
+            end_date = None
+        elif self.operator == '>':
+            start_date = None
+            end_date = today - datetime.timedelta(days=days+1)
```
Comment (dimagimon) [inline corehq/apps/app_manager/models.py]: E226 missing whitespace around arithmetic operator

**Comment by czue** [comment]
nice! guessing @nickpell should review.

**Comment by czue** [comment]
or maybe also @dannyroberts since he was just in here

File: corehq/apps/app_manager/models.py [hunk 3]
```diff
@@ -3340,6 +3341,53 @@ def get_filter_value(self, user):
...
+            '<=',
+            '>=',
+            '>',
+            '<',
+            'between'
+        ],
+        required=True,
+    )
+    date_number = StringProperty(required=True)
+    date_number2 = StringProperty()
```
Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: nitpick: could be named better (same for `days2` below)

Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: eh, nm, forgot that `date_number2` isn't necessarily the later date, maybe there isn't a much better name for it

File: corehq/apps/app_manager/templates/app_manager/partials/filter_configs.html [hunk 4]
```diff
@@ -17,37 +17,51 @@
...
+                    <div data-bind="visible: filter.selectedValue.doc_type() === 'CustomDatespanFilter'">
+                        <span>today() - </span><span data-bind="text: filter.display"></span>
+                        <select style="width:85px;"
+                                data-bind="
+                                    options: report.filterConfig.date_operators,
+                                    value: filter.selectedValue.operator
+                        "></select>
+                        <input type="number" style="width:45px;" data-bind="value: filter.selectedValue.date_number"/>
+                        <span data-bind="visible: filter.selectedValue.operator() === 'between'">and</span>
+                        <input type="number" style="width:45px;" data-bind="
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/filter_configs.html]: class `input-mini` will shrink this for you without inline styling: http://getbootstrap.com/2.3.2/base-css.html#forms

File: corehq/apps/app_manager/templates/app_manager/partials/filter_configs.html [hunk 5]
```diff
@@ -17,37 +17,51 @@
...
+                    </div>
+                    <div data-bind="visible: filter.selectedValue.doc_type() === 'StaticDatespanFilter'">
+                        <select data-bind="
+                            options: report.filterConfig.date_range_options,
+                            value: filter.selectedValue.date_range
+                        "></select>
+                    </div>
+                    <div data-bind="visible: filter.selectedValue.doc_type() === 'CustomDatespanFilter'">
+                        <span>today() - </span><span data-bind="text: filter.display"></span>
+                        <select style="width:85px;"
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/filter_configs.html]: maybe `class="span3"` instead?

File: corehq/apps/app_manager/templates/app_manager/partials/filter_configs.html [hunk 6]
```diff
@@ -17,37 +17,51 @@
...
+                            visible: filter.selectedValue.operator() === 'between'
+                        "/>
+                    </div>
+                    <div data-bind="visible: filter.selectedValue.doc_type() === 'CustomDataAutoFilter'">
+                        <input type="text" data-bind="value: filter.selectedValue.custom_data_property"/>
+                    </div>
+                    <div data-bind="visible: filter.selectedValue.doc_type() === 'StaticChoiceListFilter'">
+                        <input type="text"
+                               class="choice_filter"
+                               style="width:300px"
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/filter_configs.html]: is there a reason this is extra wide? i know it was already this big & you just moved it, just wondering.

Comment (calellowitz) [inline corehq/apps/app_manager/templates/app_manager/partials/filter_configs.html]: i am not sure, but my guess is that is it to accommodate having multiple choices selected, and being able to see them all at once in the input. i think its the only existing filter that isn't a select which may be why its wider than the others

File: corehq/apps/app_manager/templates/app_manager/partials/filter_configs.html [hunk 7]
```diff
@@ -17,37 +17,51 @@
...
-                    <input type="hidden" data-bind="value: dynamicFilterName">
-                </div>
-                <div data-bind="visible: filter.selectedValue.doc_type() === 'StaticChoiceFilter'">
-                    <select data-bind="
-                        options: filter.choices,
-                        optionsText: 'display',
-                        optionsValue: 'value',
-                        value: filter.selectedValue.select_value
-                    "></select>
+                <div style="padding-top:10px">
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/filter_configs.html]: I poked at your branch for a while trying to find a way to get rid of this wrapper div. Turns out this whole reports area is inside of a `form-horizontal`, which kills the `margin-bottom` that inputs and selects normally get. The reports area should be on a separate tab from the module settings, but that's not really within the scope of this PR. Inline styling is bad but does seem like the best option here.

**Comment by orangejenny** [comment]
I'm only commenting on styling.

The UI for this looks a good deal better than it did a couple of iterations ago, nice job on that.

File: corehq/apps/app_manager/models.py (hunk 1 above)
Comment (nickpell) [inline corehq/apps/app_manager/models.py]: these are good things to fix so there's consistency across the codebase

**Comment by nickpell** [comment]
No blocking feedback from me; going to merge this PR and let @calellowitz make any followup changes in subsequent PRs

**Pull request merged**